### PR TITLE
[Dynamo] Refactor tensor methods handling

### PR DESCRIPTION
**Pull request by oulgen**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #119631
* #119615
* __->__ #119581

Fixes part of #119128

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

File: torch/_dynamo/variables/tensor.py [hunk 1]
```diff
@@ -222,35 +222,56 @@ def var_getattr(self, tx, name):
...
-        elif name == "shape" and self.size is None:
-            result = self.call_method(tx, "size", [], {})
-        elif name == "ndim" and self.ndim is None:
-            result = self.call_method(tx, "dim", [], {})
-        elif name == "data":
-            result = self.call_method(tx, "detach", [], {})
         if name == "__class__":
             return UserDefinedClassVariable(self.python_type())
 
+        def method_ndim():
```
Comment (oulgen) [inline torch/_dynamo/variables/tensor.py:228]: I do worry that refactors like this will make the code harder to read/debug

**Review by jansel — CHANGES_REQUESTED** [review]
You shouldn't declare the methods in local scope, since it will still be O(n).  It will get mapped to a series on `N` [MAKE_FUNCTION](https://docs.python.org/3/library/dis.html#opcode-MAKE_FUNCTION) calls.

It should be on the class or global scope.

**Comment by oulgen** [comment]
@jansel Good catch, Python does not actually lift nested functions/closures.

Updated the naming schema to avoid conflicts.

**Review by jansel — APPROVED** [review]

**Review by anijain2305 — APPROVED** [review]

File: torch/_dynamo/variables/tensor.py (hunk 1 above)
Comment (anijain2305) [reply torch/_dynamo/variables/tensor.py]: I think this is ok. We have similar handler for builtins. So it should improve readability in the long run.

**Pull request closed without merging**